src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,983 @@
+      (s, n) =>
+        val range = n._1
+        s + DotStruct(s"Global_${range.start}_${range.end}", s"Global_$range", None)
+    }
+
+    structs ++= stackMapping.foldLeft(Set[DotStruct]()) {
+      (s, n) =>
+        val offset = n._1
+        s + DotStruct(s"Stack_$offset", s"Stack_$offset", None)
+    }
+
+    var arrows = nodes.foldLeft(Set[StructArrow]()) {
+      (s, node) =>
+        s ++ node.cells.foldLeft(Set[StructArrow]()) {
+          (k, c) =>
+            val offset = c._1
+            val cell = c._2
+            if cell._pointee.isDefined then
+              val pointee = find(cell.getPointee)
+              s + StructArrow(DotStructElement(node.id.toString, Some(offset.toString)), DotStructElement(pointee.node.id.toString, Some(pointee.cell.offset.toString)), pointee.internalOffset.toString)
+            else
+              s
+        }
+    }
+
+
+    arrows ++= formals.foldLeft(Set[StructArrow]()) {
+      (s, n) =>
+        val variable = n._1.name
+        val value = find(n._2)
+        s + StructArrow(DotStructElement(s"Formal_$variable", None), DotStructElement(value.node.id.toString, Some(value.cell.offset.toString)), value.internalOffset.toString)
+    }
+
+    arrows ++= varToCell.foldLeft(Set[StructArrow]()) {
+      (s, r) =>
+        val pos = r._1
+        val mapping = r._2
+        s ++ mapping.foldLeft(Set[StructArrow]()) {
+          (k, n) =>
+            val variable = n._1.name
+            val value = find(n._2)
+            k + StructArrow(DotStructElement(s"SSA_${pos.toShortString.slice(1, 9)}_$variable", None), DotStructElement(value.node.id.toString, Some(value.cell.offset.toString)), value.internalOffset.toString)
+        }
+    }
+
+
+    arrows ++= globalMapping.foldLeft(Set[StructArrow]()) {
+      (s, n) =>
+        val range = n._1
+        val node= find(n._2.node).node
+        val offset = n._2.offset + find(n._2.node).offset
+        val cellOffset = node.getCell(offset).offset
+        val internalOffset = offset - cellOffset
+        s + StructArrow(DotStructElement(s"Global_${range.start}_${range.end}", None), DotStructElement(node.id.toString, Some(cellOffset.toString)), internalOffset.toString)
+    }
+
+    arrows ++= stackMapping.foldLeft(Set[StructArrow]()) {
+      (s, n) =>
+        val offset = n._1
+        val node = find(n._2).node
+        val nodeOffset =  find(n._2).offset
+        val cellOffset = node.getCell(nodeOffset).offset
+        val internalOffset = nodeOffset - cellOffset
+        s + StructArrow(DotStructElement(s"Stack_$offset", None), DotStructElement(node.id.toString, Some(cellOffset.toString)), internalOffset.toString)
+    }
+
+
+    StructDotGraph(proc.name, structs, arrows).toDotString
+  }
+
+
+  /**
+   * Collapses the node causing it to lose field sensitivity
+   */
+  def collapseNode(n: DSN): DSN =
+
+
+    val (term, offset) = solver.find(n.term)
+    val node: DSN = term.asInstanceOf[Derm].node
```

Comment:
If you are using asInstanceOf like this, it likely means you have not used types correctly. If the first output from solver.find is guaranteed to be a Derm, that should be reflected by the types.

Reply:
Oh I see this is an issue caused by the use of types from TIP, so it's not straight-forward to fix. Hmm